### PR TITLE
[WIP] provides a mechanism for exposing health checks from controllers

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
/kind feature

**What this PR does / why we need it**: provides a mechanism for exposing health checks from controllers.

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
/assign @deads2k 

@deads2k WDYT ?

File: pkg/controller/garbagecollector/healthz.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+type cacheSyncedCheck struct {
+	controller *GarbageCollector
+}
+
+func (c cacheSyncedCheck) Name() string {
+	return "cachesSynced"
+}
+
+func (c *cacheSyncedCheck) Check(r *http.Request) error {
```
Comment (p0lyn0mial) [inline pkg/controller/garbagecollector/healthz.go]: an example that checks if the caches were synced.

running `curl ".../healthz/controllers/garbagecollector?verbose=true"` gives 
```
[+]cachesSynced ok
healthz check passed
```

running `curl ".../healthz/controllers/garbagecollector/cachesSynced"` returns 
```
ok
```

**Comment by krmayankk** [comment]
is this exposing aliveness check for each of the controllers ?

File: pkg/controller/garbagecollector/healthz.go (hunk 1 above)
Comment (deads2k) [reply pkg/controller/garbagecollector/healthz.go]: This is a great detector, but I think it's a debug handler (first arg).  If we returned unhealthy until our caches sync, then the standby process will constantly crashloop.

File: cmd/kube-controller-manager/app/apps.go [hunk 2]
```diff
@@ -26,16 +26,17 @@ import (
...
+	"k8s.io/apiserver/pkg/server/healthz"
 	"k8s.io/client-go/util/flowcontrol"
 	"k8s.io/kubernetes/pkg/controller/daemon"
 	"k8s.io/kubernetes/pkg/controller/deployment"
 	"k8s.io/kubernetes/pkg/controller/replicaset"
 	"k8s.io/kubernetes/pkg/controller/statefulset"
 )
 
-func startDaemonSetController(ctx ControllerContext) (http.Handler, bool, error) {
+func startDaemonSetController(ctx ControllerContext) (http.Handler, []healthz.HealthzChecker, bool, error) {
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/apps.go]: I think the idea of returning a slice of health checks is a good one.

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 3]
```diff
@@ -508,6 +508,11 @@ func StartControllers(ctx ControllerContext, startSATokenController InitFunc, co
 			unsecuredMux.UnlistedHandle(basePath, http.StripPrefix(basePath, debugHandler))
 			unsecuredMux.UnlistedHandlePrefix(basePath+"/", http.StripPrefix(basePath, debugHandler))
 		}
+		if unsecuredMux != nil {
+			for _, healthzHandler := range healthzHandlers {
+				healthz.InstallPathHandler(unsecuredMux, fmt.Sprintf("/healthz/controllers/%s", controllerName), healthzHandler)
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/controllermanager.go]: Yeah, the prefix is good.

**Comment by deads2k** [comment]
@kubernetes/sig-api-machinery-pr-reviews @lavalamp @cheftako @sttts  a way to return health functions back from the individual controllers.  I think the first example missed, but in concept it is a good idea.

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 3 above)
Comment (cheftako) [reply cmd/kube-controller-manager/app/controllermanager.go]: InstallPathHandler has checks as variadic, so not sure there is any point in having this inside a for loop.

File: pkg/controller/garbagecollector/healthz.go (hunk 1 above)
Comment (cheftako) [reply pkg/controller/garbagecollector/healthz.go]: +1 Davids point. If we believe that the GC has a problem when its cache is too out of date, then we need to define what too out of date is and health check based on being outside that duration.

**Comment by cheftako** [comment]
I really like the idea of per controller health functions. However I worry this will exacerbate the existing problem we have with crash looping KCMs. Should all controllers really be able to bring down the KCM. If we are ok with that, I think we should think about education/documentation on the consequences of using this feature. I also think if we add the feature to the Kube Controller Manager we need to quickly add it to the Cloud Controller Manager.

**Comment by deads2k** [comment]
> I really like the idea of per controller health functions. However I worry this will exacerbate the existing problem we have with crash looping KCMs. Should all controllers really be able to bring down the KCM. If we are ok with that, I think we should think about education/documentation on the consequences of using this feature. I also think if we add the feature to the Kube Controller Manager we need to quickly add it to the Cloud Controller Manager.

I think that if we start by constraining the initial health checks to cases where a particular controller can no longer function properly, we can safely start adding them.  One that comes to mind:  if caches that previously sync'd cannot list/watch for X minutes.   That would be a stuck list/watch that we have observed in the past. It's a little harder to watch than this.

How about splitting this check out into the debug endpoint for the controllers and then build out the healthz for an informer?  Once we get it filled out there, we can get these two pulls to meet in the middle.

File: pkg/controller/garbagecollector/healthz.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/controller/garbagecollector/healthz.go]: ah, I see, you're absolutely right. 

would it be okay to write a generic detector that would accept multiple `InformerSynced` functions, a `wait` time after which we start to fail  and perhaps optional `resync` time for things like `gc` that needs periodic synchronisation?

Comment (deads2k) [reply pkg/controller/garbagecollector/healthz.go]: > would it be okay to write a generic detector that would accept multiple `InformerSynced` functions, a `wait` time after which we start to fail and perhaps optional `resync` time for things like `gc` that needs periodic synchronisation?

I think the informer should try to summarize its health via a function that conforms to a healthcheck.  To do that, it should have limits for number of consistent failures, minimum failures, and minimum time to report unhealthy.

Comment (lavalamp) [reply pkg/controller/garbagecollector/healthz.go]: drive by comment: you can look at the way workqueue exposes metrics for a model here.

**Comment by cheftako** [comment]
> > I really like the idea of per controller health functions. However I worry this will exacerbate the existing problem we have with crash looping KCMs. Should all controllers really be able to bring down the KCM. If we are ok with that, I think we should think about education/documentation on the consequences of using this feature. I also think if we add the feature to the Kube Controller Manager we need to quickly add it to the Cloud Controller Manager.
> 
> I think that if we start by constraining the initial health checks to cases where a particular controller can no longer function properly, we can safely start adding them. One that comes to mind: if caches that previously sync'd cannot list/watch for X minutes. That would be a stuck list/watch that we have observed in the past. It's a little harder to watch than this.
> 
That's a reallt good point. It also brings up the issue of who is really unhealthy. If its a problem in the KAS which restarting the KCM won't fix then there is no point in marking the KCM as unhealthy.

**Pull request closed without merging**